### ISSUE TITLE
DNSMonitor doesn't work with ClusterServersConfig

### ISSUE THREAD

**Issue by aoskotsky-amplify** [redisson/redisson#1213]
I noticed that Redisson isn't detecting DNS changes when using the ClusterServersConfig with AWS Elasticache Redis in cluster mode. After debugging it looks like this line is never hit https://github.com/redisson/redisson/blob/master/redisson/src/main/java/org/redisson/connection/MasterSlaveConnectionManager.java#L289

I am using a CNAME DNS record that points to the Configuration endpoint of the Redis cluster and then testing destroying/recreating the cluster and pointing the CNAME at the new configuration endpoint. 

Is this expected? What's the expected behavior here?

**Comment by wubins** [comment]
@aoskotsky-amplify  which  version of redission

**Comment by mrniko** [comment]
@aoskotsky-amplify I have solved an issue with DNS resolving in last version. Have you tried it?

**Comment by aoskotsky-amplify** [comment]
I tried it with version 3.5.7

**Comment by mrniko** [comment]
Oh, sorry cluster is not supported yet. Which Redis cloud vendor do you use?

**Comment by mrniko** [comment]
DNS resolving for Cluster hasn't been implemented. I'm going to implement it in next version

**Comment by mrniko** [comment]
@aoskotsky-amplify  I'm still thinking how to do it in best way. In Redis cluster there are no single node used to read configuration. Redisson is always trying to read cluster config from any node in cluster since every node has the same configuration. Cluster config usually use ip rather than domains. This why I decided to postpone dns resolving implementation for cluster setup. What do you think?

**Comment by aoskotsky-amplify** [comment]
AWS Elasticache provides a domain name to use. IPs can change as nodes go up and down so I need to use the domain name that Elasticache gives. 

Why not use the original node address thats provided in the config and then resolve it to a list of IPs each time it's needed? This could be cached if performance is an issue but it would be very useful if Redisson eventually picked up the new IPs when a DNS change happens for a AWS Redis cluster.

**Comment by johnou** [comment]
@mrniko does Redisson cache InetSocketAddress? see https://github.com/lettuce-io/lettuce-core/issues/144

**Comment by mrniko** [comment]
@aoskotsky-amplify I got you. I'll research that case. Thank you!

**Comment by mrniko** [comment]
@johnou no, it doesn't. all InetSocketAddress bound to client object. If client goes down or its IP has been updated then this client is shutdown.

**Comment by mrniko** [comment]
Done

**Comment by WuKongQC** [comment]
Is this problem solved?  I tryed ,but did not work,  because of netty cached dns and will not update dns relation.

**Comment by mrniko** [comment]
@WuKongQC 

Which netty version do you use? Could you provide netty dns resolver log?

**Comment by WuKongQC** [comment]
version info     redisson:3.13.6      netty: 4.1.52.Final   

redis config:
redisson:
    # 集群节点，格式：redis://ip:port
    address:
      - redis://dcluster-redis:6380

first, set hosts
  172.28.56.237 dcluster-redis

 redisson log:
[DEBUG] 2020-10-15 16:53:50.125 [] [] [nioEventLoopGroup-5-5] ClusterConnectionManager.lambda$updateClusterState$10[453]:cluster nodes state got from /172.28.56.237:6380:

than, changed hosts
  172.28.19.80 dcluster-redis

in redisson log, it use old IP  :
[DEBUG] 2020-10-15 16:55:20.503 [] [] [nioEventLoopGroup-5-5] ClusterConnectionManager.lambda$updateClusterState$10[453]:cluster nodes state got from /172.28.56.237:6380:

no netty log, I debug into netty and find netty used cached dns relation.
dns code in class:  io.netty.resolver.dns.DnsNameResolver  doResolveAll

**Comment by mrniko** [comment]
NoopDnsCnameCache.INSTANCE is used by default as resolveCache instance

**Comment by mrniko** [comment]
You need to adjust TTL returned by the DNS server

**Comment by WuKongQC** [comment]
not through DNS server,   I set domain to  ip relation in operate system hosts file .
I added one line code as follow, when create RedissonClient Bean.   solved this problem. 
//class org.redisson.config
config.setAddressResolverGroupFactory((var1, var2)->DefaultAddressResolverGroup.INSTANCE);

**Comment by johnou** [comment]
@WuKongQC do you use a security manager? by default if you set one, java will ignore the ttl and return the first resolved ip(s). To work around this you can set the ttl when bootstrapping your application but it must happen before the first socket connection..

```java
public class Main {

    static {
        java.security.Security.setProperty("networkaddress.cache.ttl", "10");
        java.security.Security.setProperty("networkaddress.cache.negative.ttl", "10");
    }

    ...
}
```

https://docs.oracle.com/javase/7/docs/technotes/guides/net/properties.html

networkaddress.cache.ttl
Specified in java.security to indicate the caching policy for successful name lookups from the name service.. The value is specified as integer to indicate the number of seconds to cache the successful lookup.
A value of -1 indicates "cache forever". **The default behavior is to cache forever when a security manager is installed, and to cache for an implementation specific period of time, when a security manager is not installed.**

**Comment by WuKongQC** [comment]
I have tryed this method, does not work.
the root cause is that netty bootstrap's dns cache never expire, that is what i saw in netty source code.

**Comment by mrniko** [comment]
@WuKongQC 

Probably it's correct behaviour for dns entries stored in file. I would recommend to use MaraDNS instead https://maradns.samiam.org/

**Comment by xiankris156** [comment]
@WuKongQC I still can't switch to the new IP via the domain name even after using config.setAddressResolverGroupFactory((var1, var2) -> DefaultAddressResolverGroup.INSTANCE). Here is my bean code.
@Bean(destroyMethod = "shutdown")
    public RedissonClient redissonClient() {
        Config config = new Config();
        
        // 解析集群节点
        List<String> nodes = Arrays.asList(clusterNodes.split(","));
        
        // 配置集群模式
        ClusterServersConfig clusterConfig = config.useClusterServers()
                .addNodeAddress(nodes.stream().map(node -> "redis://" + node).toArray(String[]::new))
                .setScanInterval(scanInterval) // 集群节点扫描间隔(毫秒)，用于拓扑感知
                .setConnectTimeout(timeout)
                .setRetryAttempts(60000)
                .setDnsMonitoringInterval(3000)
                .setTimeout(timeout);
        
        // 设置密码
        if (password != null && !password.isEmpty()) {
            clusterConfig.setPassword(password);
        }
        config.setAddressResolverGroupFactory((var1, var2)-> DefaultAddressResolverGroup.INSTANCE);
      
        
        return Redisson.create(config);
    }
